### ISSUE TITLE
Implement Training Loop (Layer 8)

### ISSUE THREAD

**Issue by noahgift** [paiml/entrenar#4]
# Layer 8: High-Level Training Loop

Implement the Trainer abstraction that ties together model, optimizer, and loss function into a cohesive training API.

## Features
- Trainer struct with model, optimizer, loss_fn, context
- train_step() - Single batch training step
- train_epoch() - Full epoch training
- Support for gradient clipping
- Support for LR scheduling
- Checkpoint saving
- Progress tracking

## Spec
```rust
pub struct Trainer<'g> {
    model: Model<'g>,
    optimizer: Box<dyn Optimizer>,
    loss_fn: Box<dyn LossFn>,
    ctx: Context<'g>,
}

impl<'g> Trainer<'g> {
    pub fn train_step(&mut self, batch: &Batch) -> f32;
    pub fn train_epoch(&mut self, dataloader: &DataLoader) -> f32;
}
```

## Acceptance Criteria
- ✅ Trainer struct with generic model/optimizer/loss
- ✅ train_step with forward/backward/update
- ✅ train_epoch with epoch-level metrics
- ✅ Gradient clipping support
- ✅ LR scheduler integration
- ✅ Comprehensive tests
- ✅ Example training loop

Estimated: 32h
